### ISSUE TITLE
Can't understand WAIT_FOR_DATA and WAIT_FOR_OTHER_TENSOR_DATA

### ISSUE THREAD

**Issue by yogurfrul** [horovod/horovod#442]
I run horovod.tensorflow , and we know gradient is the input to allreduce op  ,they are data dependencies. So when the HorovodAllreduce op start must after the gradient computing op.
How can Tensorflow smartly interleave scheduling and GPU computation to make HorovodAllreduce start before gradient compute ready!
And I read the source code , I get the WAIT_FOR_DATA sub-phases just check the ready_event save in the state_table ,
is  it necessary to check ready_event ?
While  I found this in timeline [readme](https://github.com/uber/horovod/blob/master/docs/timeline.md)
```
**WAIT_FOR_DATA** 
indicates time taken to wait for GPU to finish computing input to the allreduce, allgather, or broadcast operations. This happens because TensorFlow tries to smartly interleave scheduling and GPU computation. This is only applicable to situations where the Horovod operation is placed on GPU.  
**WAIT_FOR_OTHER_TENSOR_DATA**
 indicates time taken to wait for GPU to finish computing other inputs for other operations that are part of the same fusion batch.
```

**Comment by anpark** [comment]
this is for fusion tensor data check, not for HorovodAllreduce start before gradient compute ready.

HorovodAllreduce start before gradient compute ready checked by ncclallgather or mpi to sync

**Comment by yogurfrul** [comment]
I delete the wait_for_data code in   operations.cc from line 754 to line 779 .https://github.com/uber/horovod/blob/master/horovod/common/operations.cc#L754
and reinstall horovod
it run ok

**Comment by yogurfrul** [comment]
check what ?
all the gradients would check readiness before pushed into ready_to_reduce vector.

**Comment by yogurfrul** [comment]
I found this in tensoeflow/mpi_ops.cc
when the HorovodAllreduceOp start ,we know the data is ready,but output is allocated async,
so ,if we don't check the ReadyEvent before actully run ncclAllreduce operation  in PerformOperation func, maybe when ncclallreduce finished the output is still not allocated ,memcpy out from fusion buffer to output would get an andress error?
@alsrgv @anpark it is right?
But I still cannot find where the status of  ReadyEvent changed to Ready( not Pedding in code)
```
    OP_REQUIRES_OK_ASYNC(
        context, context->allocate_output(0, tensor.shape(), &output), done);
    // ReadyEvent makes sure input tensor is ready, and output is allocated.
    auto ready_event = std::shared_ptr<common::ReadyEvent>(RecordReadyEvent(context));
```

**Comment by alsrgv** [comment]
@yogurfrul, what we found is that both TensorFlow and PyTorch will start running Horovod op **before** GPU kernels are fully completed.  That's because GPU kernels are usually asynchronous, and so simply executing ops like this:
```
1. cuda_conv_kernel(..., stream);
2. mpi_allreduce();
```

Will potentially cause mpi_allreduce to perform the operation on garbage.  For that reason, we add event synchronization between (1) and (2).
